File: DSA-PYTHON/Binary_Search.py
# def search (arr, s, e, target) -> bool:
#   arr.sort()

#   mid = int(s + (e - s)) // 2

#   if (arr[mid] == target):
#     return True

#   elif (arr[mid] < target):
#     return search(arr, mid + 1, e, target)
#   else:
#     return search(arr, s, mid - 1, target)

# print(search([1, 2, 3, 4, 5], 0, 4, 1))

# def reverse(ans, word):
#   l = len(word)

#   if (l == 0):
#     return ans

#   return ans + word[l - 1] + reverse(ans, word[:l - 1])

# print(reverse("", 'abcd'))


# def sort(lst, e):

#   if (e <= 1):
#     return lst
  
#   for i in range(e):
#     if (i + 1 < e and lst[i] > lst[i + 1]):
#       lst[i], lst[i + 1] = lst[i + 1], lst[i]
  
#   return sort(lst, e - 1)

# print(sort([5, 4, 3, 2, 1], 5))

# from typing import List

# class Solution:
#     def sub_sets(self, arr, ans, i):
#         if i >= len(arr):
#             return [ans]  # Wrap ans in a list and return it

#         num = arr[i]

#         # Include the current element
#         included = self.sub_sets(arr, ans + [num], i + 1)

#         # Not include the current element
#         not_included = self.sub_sets(arr, ans, i + 1)

#         return included + not_included  # Combine the two lists

#     def subsets(self, arr: List[int]) -> List[List[int]]:
#         return self.sub_sets(arr, [], 0)

# # Usage
# solution = Solution()
# result = solution.subsets([1, 2, 3])
# print(result)


def isSubsequence(s: str, t: str) -> bool:
        while t:
            if s[0] == t[0]:
                s = s[1:]
                if s == "":
                  return True
            
            t = t[1:]
        return False

print(isSubsequence('abc', 'ahbgdc'))

File: DSA/SET - 1/VERSION-1/1/1VERSION-BETA/DAY1-DAY17/com/mithin/BinarySearch.java
package com.mithin;

public class BinarySearch {
    public static void main(String[] args) {
        int[] arr = {2, 4, 6, 9, 11, 12, 14, 20, 36, 48};
        int target = 36;
        int ans = BinarySh(arr, target);
        System.out.println(ans);

    }
    static int BinarySh(int[] arr, int target){
        if (arr.length == 0){
            return -1;
        }
        int start = 0;
        int end = arr.length-1;
//        int mid = start + (end-start)/2;
        while(start <= end){
            int mid = start + (end-start)/2;
            if(target == arr[mid]){
                return mid;
            }
            if(target > arr[mid]){
                start = mid + 1;
                if(arr[mid] == target){
                    return mid;
                }
            }
            else{
                end = mid-1;
                if (target == arr[mid]){
                    return mid;
                }
            }
        }
        return -1;
    }

}
